File: backend/app/utils/requirement_normalization.py
"""Normalize requirement review inputs into a canonical review representation."""

from __future__ import annotations

import re
import unicodedata

from app.models.review_models import RequirementReviewInput

_FIELD_LABELS: tuple[tuple[str, str], ...] = (
    ("project id", "project_id"),
    ("global id", "global_id"),
    ("status", "status"),
    ("release", "release"),
    ("title", "title"),
    ("description", "description"),
    ("requirement volatility", "requirement_volatility"),
    ("rationale", "rationale"),
    ("fdal", "fdal"),
    ("sal", "sal"),
    ("derived requirement", "derived_requirement"),
    ("safety requirement", "safety_requirement"),
    ("security effectiveness requirement", "security_effectiveness_requirement"),
    ("validation method", "validation_method"),
    ("verification method", "verification_method"),
    ("assigned to", "assigned_to"),
    ("reference information", "reference_information"),
    ("created by", "created_by"),
    ("created date", "created_date"),
    ("modified by", "modified_by"),
    ("modified date", "modified_date"),
    ("last activity date", "last_activity_date"),
)


def normalize_requirement_review_input(payload: RequirementReviewInput) -> RequirementReviewInput:
    """Return a canonicalized copy of a review input.

    Fielded document text is reduced to the actual requirement statement, while
    supporting fields are preserved in metadata so pasted text and extracted
    document text review against the same semantic input.
    """
    normalized_text, extracted_metadata = _canonicalize_text(payload.text)
    merged_metadata = dict(payload.metadata)
    merged_metadata.update(extracted_metadata)

    requirement_id = payload.requirement_id or merged_metadata.get("project_id")
    if requirement_id:
        requirement_id = requirement_id.strip()

    updates: dict[str, object] = {
        "text": normalized_text,
        "metadata": merged_metadata,
    }
    if requirement_id:
        updates["requirement_id"] = requirement_id

    return payload.model_copy(update=updates)


def _canonicalize_text(text: str) -> tuple[str, dict[str, str]]:
    cleaned_lines = [_normalize_line(line) for line in _split_lines(text)]
    extracted: dict[str, str] = {}

    current_label: str | None = None
    current_value_parts: list[str] = []
    description_parts: list[str] = []
    description_active = False

    index = 0
    while index < len(cleaned_lines):
        line = cleaned_lines[index]
        if not line:
            index += 1
            continue

        label, span, value = _consume_label(cleaned_lines, index)
        if label is not None:
            if current_label is not None:
                _store_field(extracted, current_label, current_value_parts)
            current_label = label
            current_value_parts = []
            description_active = label == "description"
            if value:
                current_value_parts.append(value)
                if description_active:
                    description_parts.append(value)
            index += span
            continue

        if current_label is not None:
            current_value_parts.append(line)
            if description_active:
                description_parts.append(line)
        index += 1

    if current_label is not None:
        _store_field(extracted, current_label, current_value_parts)

    if description_parts:
        canonical_text = " ".join(" ".join(description_parts).split())
        return canonical_text, extracted

    return " ".join(" ".join(cleaned_lines).split()), extracted


def _split_lines(text: str) -> list[str]:
    normalized = unicodedata.normalize("NFKC", text)
    normalized = normalized.replace("\r\n", "\n").replace("\r", "\n")
    normalized = "".join(ch for ch in normalized if ch == "\n" or ch == "\t" or not unicodedata.category(ch).startswith("C"))
    return normalized.split("\n")


def _normalize_line(line: str) -> str:
    line = re.sub(r"\s+", " ", line.replace("\t", " ")).strip()
    line = re.sub(r"^\s*[-*•]+\s*", "", line)
    line = _strip_markdown_links(line)
    line = re.sub(r"^\*{1,2}", "", line)
    line = re.sub(r"\*{1,2}$", "", line)
    return line.strip()


def _strip_markdown_links(line: str) -> str:
    return re.sub(r"\[([^\]]+)\]\([^)]+\)", r"\1", line)


def _consume_label(lines: list[str], index: int) -> tuple[str | None, int, str]:
    max_width = min(4, len(lines) - index)
    for width in range(max_width, 0, -1):
        candidate_parts = [part for part in lines[index : index + width] if part]
        if not candidate_parts:
            continue

        candidate = " ".join(candidate_parts)
        lowered = candidate.lower().rstrip(":")
        for label, field in _FIELD_LABELS:
            if lowered == label:
                return field, width, ""
            if lowered.startswith(f"{label} "):
                remainder = candidate[len(label):].lstrip(" :\t-")
                return field, width, remainder.strip()
    return None, 0, ""


def _store_field(extracted: dict[str, str], field: str, parts: list[str]) -> None:
    value = " ".join(" ".join(parts).split()).strip()
    if value:
        extracted[field] = value

File: frontend/src/radia_ai/features/jamaRequirementReviewer/utils/requirementNormalization.ts
/**
 * Client-side port of the backend requirement normalization logic.
 *
 * Given PDF-extracted text from a Jama export, extracts the key fields
 * (Title, Description/body, Rationale) and returns a clean structured string
 * that is shown in the UI preview and sent to the AI for review.
 */

/** Known Jama metadata field labels, longest first for greedy matching. */
const FIELD_LABELS: [label: string, field: string][] = [
  ['security effectiveness requirement', 'security_effectiveness_requirement'],
  ['requirement volatility', 'requirement_volatility'],
  ['derived requirement', 'derived_requirement'],
  ['safety requirement', 'safety_requirement'],
  ['last activity date', 'last_activity_date'],
  ['verification method', 'verification_method'],
  ['validation method', 'validation_method'],
  ['reference information', 'reference_information'],
  ['modified date', 'modified_date'],
  ['created date', 'created_date'],
  ['modified by', 'modified_by'],
  ['created by', 'created_by'],
  ['assigned to', 'assigned_to'],
  ['global id', 'global_id'],
  ['project id', 'project_id'],
  ['description', 'description'],
  ['rationale', 'rationale'],
  ['release', 'release'],
  ['status', 'status'],
  ['title', 'title'],
  ['fdal', 'fdal'],
  ['sal', 'sal'],
  ['rev', 'release'],
];

const CORE_FIELD_LABELS = new Set(['title', 'description', 'rationale']);
const TRAILING_METADATA_LABELS = FIELD_LABELS.map(([label]) => label).filter((label) => !CORE_FIELD_LABELS.has(label));

function stripMarkdownLinks(line: string): string {
  return line.replace(/\[([^\]]+)\]\([^)]+\)/g, '$1');
}

function normalizeLine(line: string): string {
  let l = line.replace(/\t/g, ' ').replace(/\s+/g, ' ').trim();
  l = l.replace(/^\s*[-*•]+\s*/, '');
  l = stripMarkdownLinks(l);
  l = l.replace(/^\*{1,2}/, '').replace(/\*{1,2}$/, '');
  return l.trim();
}

/** Returns [fieldKey, inlineValue] if the line starts with a known field label, else null. */
function matchFieldLabel(line: string): [string, string] | null {
  const lowered = line.toLowerCase().replace(/:$/, '').trim();
  for (const [label, field] of FIELD_LABELS) {
    if (lowered === label) return [field, ''];
    if (lowered.startsWith(label + ' ') || lowered.startsWith(label + ':')) {
      const remainder = line.slice(label.length).replace(/^[\s:\t-]+/, '').trim();
      return [field, remainder];
    }
  }
  return null;
}

function escapeRegex(text: string): string {
  return text.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function truncateAtTrailingMetadata(line: string): string {
  const lowered = line.toLowerCase();
  let cutoff = line.length;
  for (const label of TRAILING_METADATA_LABELS) {
    const regex = new RegExp(`\\b${escapeRegex(label)}\\b`, 'i');
    const match = regex.exec(lowered);
    if (match && match.index < cutoff) {
      cutoff = match.index;
    }
  }
  return line.slice(0, cutoff).trimEnd();
}

function containsTrailingMetadata(line: string): boolean {
  return TRAILING_METADATA_LABELS.some((label) => {
    const regex = new RegExp(`\\b${escapeRegex(label)}\\b`, 'i');
    return regex.test(line);
  });
}

/** True if the line looks like a Jama section heading: "1 WR-ACR-732 Some Title" */
function isSectionHeading(line: string): boolean {
  return /^\d+\s+[A-Z]{2,}-[A-Z]+-\d+\b/.test(line);
}

function extractHeadingTitle(line: string): string {
  return line.replace(/^\d+\s+[A-Z]{2,}-[A-Z]+-\d+\s+/, '').trim();
}

/**
 * Parse a Jama PDF export into its key fields.
 *
 * Returns the requirement body (text before the metadata table), title, and rationale
 * as separate strings. The body maps to what Jama calls "Description".
 */
function extractFields(raw: string): { body: string; title: string; rationale: string } {
  const lines = raw
    .replace(/\r\n?/g, '\n')
    .split('\n')
    .map(normalizeLine);

  const bodyParts: string[] = [];
  const fields: Record<string, string[]> = {};
  let currentField: string | null = null;
  let seenFields = false;
  let stopParsing = false;
  let headingTitle = '';

  for (const line of lines) {
    if (!line) continue;
    if (stopParsing) break;
    if (isSectionHeading(line)) {
      if (!headingTitle) headingTitle = extractHeadingTitle(line);
      continue;
    }

    const match = matchFieldLabel(line);
    if (match !== null) {
      const [fieldKey, inlineValue] = match;
      seenFields = true;
      currentField = fieldKey;
      if (!fields[fieldKey]) fields[fieldKey] = [];
      const content = truncateAtTrailingMetadata(inlineValue);
      if (content) fields[fieldKey].push(content);
      if (containsTrailingMetadata(inlineValue)) stopParsing = true;
      continue;
    }

    if (!seenFields) {
      bodyParts.push(line);
    } else if (currentField !== null) {
      const content = truncateAtTrailingMetadata(line);
      if (content) {
        if (!fields[currentField]) fields[currentField] = [];
        fields[currentField].push(content);
      }
      if (containsTrailingMetadata(line)) {
        stopParsing = true;
      }
    }
  }

  const joinField = (key: string) =>
    (fields[key] ?? []).join(' ').replace(/\s+/g, ' ').trim();

  // "description" field wins over the pre-field body if both exist
  const body = joinField('description') || bodyParts.join(' ').replace(/\s+/g, ' ').trim();
  const title = joinField('title') || headingTitle;
  const rationale = joinField('rationale');

  return { body, title, rationale };
}

/**
 * Normalize a Jama PDF export into a structured string containing
 * Title, Description, and Rationale — the fields the AI uses for review.
 *
 * Falls back to the full cleaned text when no structure is detected.
 */
export function normalizeRequirementText(raw: string): string {
  const { body, title, rationale } = extractFields(raw);

  // If no structure was found, return the raw text as-is
  if (!body && !title && !rationale) {
    return raw
      .replace(/\r\n?/g, '\n')
      .split('\n')
      .map(normalizeLine)
      .filter(Boolean)
      .join(' ')
      .replace(/\s+/g, ' ')
      .trim();
  }

  const parts: string[] = [];
  if (title) parts.push(`Title: ${title}`);
  if (body) parts.push(`Description: ${body}`);
  if (rationale) parts.push(`Rationale: ${rationale}`);

  return parts.join('\n\n');
}
